option.go

```diff
@@ -325,13 +330,74 @@ func (optSet *OptionSet) SetDefaults() error {
 			)
 			continue
 		}
-		(*optSet)[i].ValueSource = ValueSourceDefault
-		if err := opt.Value.Set(opt.Default); err != nil {
+		groupByValue[opt.Value] = append(groupByValue[opt.Value], opt)
+	}
+
+	for _, opts := range groupByValue {
+		// Sort the options by priority and whether or not a default is
+		// set. This won't affect the value but represents correctness
+		// from whence the value originated.
+		slices.SortFunc(opts, func(a, b *Option) int {
+			if a.ValueSource != b.ValueSource {
+				for _, vs := range valueSourcePriority {
+					if a.ValueSource == vs {
+						return -1
+					}
+					if b.ValueSource == vs {
+						return 1
+					}
+				}
+			}
+			if a.Default != b.Default {
+				if a.Default == "" {
+					return 1
+				}
+				if b.Default == "" {
+					return -1
+				}
+			}
+			return 0
+		})
+
+		// If the first option has a value source, then we don't need to
+		// set the default, but mark the source for all options.
+		if opts[0].ValueSource != ValueSourceNone {
+			for _, opt := range opts[1:] {
+				opt.ValueSource = opts[0].ValueSource
+			}
+			continue
+		}
+
+		var optWithDefault *Option
+		for _, opt := range opts {
+			if opt.Default == "" {
+				continue
+			}
+			if optWithDefault != nil && optWithDefault.Default != opt.Default {
+				merr = multierror.Append(
+					merr,
+					xerrors.Errorf(
+						"parse %q: multiple defaults set for the same value: %q and %q (%q)",
+						opt.Name, opt.Default, optWithDefault.Default, optWithDefault.Name,
+					),
+				)
+				continue
+			}
+			optWithDefault = opt
```

Comment:
suggestion: this also looks like it could be extracted to its own function and tested independently

Reply:
I don't see a benefit currently as this is pretty much "the meat" of this function and we don't need to call it elsewhere.